processor/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,504 @@
+		dpCalls.SetIntVal(c)
+
+		dimensions, ok := p.dimensionsForSeries(key)
+		if !ok {
+			return fmt.Errorf("failed to find dimensions for key %s", key)
+		}
+
+		dimensions.CopyTo(dpCalls.Attributes())
+	}
+
+	for key, c := range p.reqFailedTotal {
+		mCount := ilm.Metrics().AppendEmpty()
+		mCount.SetDataType(pmetric.MetricDataTypeSum)
+		mCount.SetName("request_failed_total")
+		mCount.Sum().SetIsMonotonic(true)
+		// TODO: Support other aggregation temporalities
+		mCount.Sum().SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
+
+		dpCalls := mCount.Sum().DataPoints().AppendEmpty()
+		dpCalls.SetStartTimestamp(pcommon.NewTimestampFromTime(p.startTime))
+		dpCalls.SetTimestamp(pcommon.NewTimestampFromTime(time.Now()))
+		dpCalls.SetIntVal(c)
+
+		dimensions, ok := p.dimensionsForSeries(key)
+		if !ok {
+			return fmt.Errorf("failed to find dimensions for key %s", key)
+		}
+
+		dimensions.CopyTo(dpCalls.Attributes())
+	}
+
+	return nil
+}
+
+func (p *processor) collectLatencyMetrics(ilm pmetric.ScopeMetrics) error {
+	for key := range p.reqDurationSecondsCount {
+		mDuration := ilm.Metrics().AppendEmpty()
+		mDuration.SetDataType(pmetric.MetricDataTypeHistogram)
+		mDuration.SetName("request_duration_seconds")
+		// TODO: Support other aggregation temporalities
+		mDuration.Histogram().SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
+
+		timestamp := pcommon.NewTimestampFromTime(time.Now())
+
+		dpDuration := mDuration.Histogram().DataPoints().AppendEmpty()
+		dpDuration.SetStartTimestamp(pcommon.NewTimestampFromTime(p.startTime))
+		dpDuration.SetTimestamp(timestamp)
+		dpDuration.SetExplicitBounds(pcommon.NewImmutableFloat64Slice(p.reqDurationBounds))
+		dpDuration.SetBucketCounts(pcommon.NewImmutableUInt64Slice(p.reqDurationSecondsBucketCounts[key]))
+		dpDuration.SetCount(p.reqDurationSecondsCount[key])
+		dpDuration.SetSum(p.reqDurationSecondsSum[key])
+
+		// TODO: Support exemplars
+
+		dimensions, ok := p.dimensionsForSeries(key)
+		if !ok {
+			return fmt.Errorf("failed to find dimensions for key %s", key)
+		}
+
+		dimensions.CopyTo(dpDuration.Attributes())
+	}
+	return nil
+}
+
+func (p *processor) buildMetricKey(clientName, serverName, connectionType string, edgeDimensions map[string]string) string {
+	var metricKey strings.Builder
+	metricKey.WriteString(clientName + metricKeySeparator + serverName + metricKeySeparator + connectionType)
+
+	for _, dimName := range p.config.Dimensions {
+		dim, ok := edgeDimensions[dimName]
+		if !ok {
+			continue
+		}
+		metricKey.WriteString(metricKeySeparator + dim)
+	}
+
+	return metricKey.String()
+}
+
+// cacheLoop runs in a separate goroutine and periodically cleans the cache
+func (p *processor) cacheLoop(d time.Duration) {
+	t := time.NewTicker(d)
+	for {
+		select {
+		case <-t.C:
+			p.cleanCache()
+		case <-p.shutdownCh:
+			return
+		}
+	}
+
+}
+
+// cleanCache removes series that have not been updated in 15 minutes
```

Comment:
Isn't there a cache TTL already specified somewhere? Should that be used instead?

Reply:
These are two different configs. The TTL you're mentioning is tracking how long edges are kept in the store before they're recorded even if incomplete. This cache cleanup is removing stale metrics that haven't received an update in 15 minutes.